Move semicolon hack into lexer

nik9000 opened the pull request:
Perviously we used token level lookbehind in the parser. That worked,
but only if the parser didn't have any ambiguity *at all*. Since the
parser has ambiguity it didn't work everywhere. In particular it failed
when parsing blocks in lambdas like `a -> {int b = a + 2; b * b}`.

This moves the hack from the parser into the lexer. There we can use
token lookbehind (same trick) to *insert* semicolons into the token
stream. This works much better for antlr because antlr's prediction
code can work with real tokens.

Also, the lexer is simpler than the parser, so if there is a place
to introduce a hack, that is a better place.

nik9000 commented:
@rmuir this fixes `;`s inside of lambdas.

@jdconrad, this is for you to review. Antlr likes this way better because it doesn't have anything tricky in the parser grammar.

jdconrad commented on modules/lang-painless/src/main/antlr/PainlessParser.g4:
```diff
@@ -91,19 +91,7 @@ trap
-            RBRACK
-        { _input.seek(index); _input.release(mark); }
-    ;
+    | EOF;
```
Nit: please drop the semicolon down a line to conform to the style of the other multi-line rules.

jdconrad commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/SuperPainlessLexer.java:
```diff
@@ -0,0 +1,96 @@
+ * ambiguity we hack them here where we can use heuristics to do it quickly.
+ * </ul>
+ */
+final class SuperPainlessLexer extends PainlessLexer {
```
I don't have a naming suggestion, but when I think of super I immediately think of super class.  Any thoughts on a different name?

nik9000 commented:
Thanks for reviewing @jdconrad ! I renamed the class, fixed the style and merged. We can rename the class again if anyone comes up with a better name.
